test/lib/middleware.test.js

```diff
@@ -0,0 +1,224 @@
+          done()
+        })
+    })
+
+    it('responds with 500 if the log throws an error', done => {
+      const originalInfo = mockLogger.info
+      mockLogger.info = () => { throw new Error('some error') }
+      request(app)
+        .post('/log/info')
+        .send({ message })
+        .expect(500, () => {
+          mockLogger.info = originalInfo
+          done()
+        })
+    })
+  })
+
+  describe('middleware', () => {
+    let status
+    beforeEach(() => {
+      status = 200
+      app.use((req, res) => res.sendStatus(status))
+    })
+
+    it('only logs the request once', done => {
+      request(app)
+        .post('/some_path')
+        .send({ message })
+        .expect(status, () => {
+          expect(mockLogger.log).to.have.been.called.exactly(1)
+          done()
+        })
+    })
+
+    it('logs in debug level if the status is a successful one', done => {
+      request(app)
+        .get('/some_path')
+        .expect(status, () => {
+          expect(mockLogger.debug).to.have.been.called()
+          expect(mockLogger.info).to.not.have.been.called()
+          expect(mockLogger.warn).to.not.have.been.called()
+          expect(mockLogger.error).to.not.have.been.called()
+          expect(mockLogger.fatal).to.not.have.been.called()
+          done()
+        })
+    })
+
+    it('logs in warn level if the status is client fault', done => {
+      status = 400
+      request(app)
+        .get('/some_path')
+        .expect(status, () => {
+          expect(mockLogger.debug).to.not.have.been.called()
+          expect(mockLogger.info).to.not.have.been.called()
+          expect(mockLogger.warn).to.have.been.called()
+          expect(mockLogger.error).to.not.have.been.called()
+          expect(mockLogger.fatal).to.not.have.been.called()
+          done()
+        })
+    })
+
+    it('logs in error level if the status is server fault', done => {
+      status = 500
+      request(app)
+        .get('/some_path')
+        .expect(status, () => {
+          expect(mockLogger.debug).to.not.have.been.called()
+          expect(mockLogger.info).to.not.have.been.called()
+          expect(mockLogger.warn).to.not.have.been.called()
+          expect(mockLogger.error).to.have.been.called()
+          expect(mockLogger.fatal).to.not.have.been.called()
+          done()
+        })
+    })
+
+    it('only calls Logger creation once', done => {
+      const originalMockLogger = MockLogger
+      MockLogger = chai.spy(MockLogger)
+      app = express()
+      app.use((req, res, next) => {
+        req.logger = mockLogger
+        next()
+      })
+      app.use(middleware)
+      app.use((req, res) => res.sendStatus(status))
+
+      request(app)
+        .get('/some_path')
+        .expect(status, () => {
+          expect(MockLogger).to.not.have.been.called()
+          MockLogger = originalMockLogger
+          done()
+        })
+    })
+
+    describe('log message', () => {
```

Comment:
I don't really like a lot of sub-describes. Especially when we check the same function in all tests.
The problem with that is that you move part of the description of the test away from the test itself.
When you read the description of one of the test you need to remember in which of the sub-describe you are now and fill on the blank.

A better approch is to start each test with:
`logs a message contains ...`
which would be read in the tests as:
`Middleware logs a message contains ...'

Reply:
But then I would also have to repeat the contents of the `beforeEach` in each of them.

Reply:
Do this for all tests.
Or even better: why are you doing this and not using a spy?